### ISSUE TITLE
Wrong layers' updates possible with different optimizers only using type info

### ISSUE THREAD

**Issue by rahulbhalley** [tensorflow/swift-apis#656]
I thought about asking transfer learning question at Google groups but found a hack/bug along the way. So I decided to open an issue here.

I am trying to perform transfer learning on my simple model but it seems like I can't pass particular layers into `Optimizer` instance for being updated instead of whole model. The hack I found is initializing the `Optimizer<Scalar>` instance and based on the inferred `Scalar` type via `for:` argument I can update any layer of that type from same network or other network. 

Based on current implementation `for:` argument only cares about type and [not the model](https://github.com/tensorflow/swift-apis/blob/7e2212efc4b41beedb40524f64649f125a554c71/Sources/TensorFlow/Optimizers/SGD.swift#L38) (the model is not used anywhere). `Optimizer` uses Model information to perform `inout` operations [here](https://github.com/tensorflow/swift-apis/blob/7e2212efc4b41beedb40524f64649f125a554c71/Sources/TensorFlow/Optimizers/SGD.swift#L54). It's a bug! 🤔

```swift
struct MyNet: Layer {
    typealias Input = Tensor<Float>
    typealias Output = Tensor<Float>
    
    var dense1 = Dense<Float>(inputSize: 10, outputSize: 10)
    var dense2 = Dense<Float>(inputSize: 10, outputSize: 10)
    var dense3 = Dense<Float>(inputSize: 10, outputSize: 10)
    
    @differentiable
    func callAsFunction(_ input: Input) -> Output {
        input.sequenced(through: dense1, dense2, dense3)
    }
}
```
Following are two optimizers for two different layers of same network having different learning rates:
```swift
var myNet = MyNet()
var lr1: Float = 0.001
var lr2: Float = 0.1
var opt1 = SGD(for: myNet.dense1, learningRate: lr1)
var opt2 = SGD(for: myNet.dense2, learningRate: lr2)
```
I get `MyNet.TangentVector` when I compute gradients of `myNet` wrt `myNet`'s current parameters at `myNet` parameters.
```swift
var grads = gradient(at: myNet) { (myNet) -> Tensor<Float> in
    myNet(Tensor<Float>(randomNormal: [1, 10])).sum()
}
```
But now is the problem: I can update parameters of any layer of type `Dense<Float>` along _any other_ layer's gradient as long as the dimension of parameters match.

Case 1: update `dense1` along `dense1` grads with `opt1`. This is expected behavior.
```swift
opt1.update(&myNet.dense1, along: grads.dense1)
print(grads.dense1.bias)
print(myNet.dense1.bias)
```
OUTPUT
```
[-0.94615835,  0.70046747,  -0.9097381, -0.46321192, -0.22250375,  0.63690734, -0.38832983,
  -1.3212568, -0.29594737,  0.07065146]
[ 0.00094615837,  -0.0007004675,   0.0009097382,  0.00046321194,  0.00022250376,
 -0.00063690735,  0.00038832985,   0.0013212569,   0.0002959474,  -7.065146e-05]
```
Case 2: update `dense2` along `dense1` grads with `opt1` although only `opt2` should update `dense2` along `dense2` grads. This is not expected but is possible.
```swift
opt1.update(&myNet.dense2, along: grads.dense1)
print(grads.dense1.bias)
print(myNet.dense2.bias)
```
OUTPUT
```
[-0.94615835,  0.70046747,  -0.9097381, -0.46321192, -0.22250375,  0.63690734, -0.38832983,
  -1.3212568, -0.29594737,  0.07065146]
[ 0.00094615837,  -0.0007004675,   0.0009097382,  0.00046321194,  0.00022250376,
 -0.00063690735,  0.00038832985,   0.0013212569,   0.0002959474,  -7.065146e-05]
```
<s>One suggestion for fixing this problem is to allow array of layers to be updated by passing them to `for:` argument and update only those without the requirement to pass `&layers` to `update()` method on optimizer akin to PyTorch's `optimizer.step()`. </s>

<s>The call in Swift case becomes `opt.updateParameters(along: grads)` instead of `opt.update(&layer, along: grad)` where `&layer` is redundant argument because `Optimizer` initializer already captures the layer to be optimized (as per syntax).</s>

Please ignore the suggestion. This API design is good if made more concrete under the hood.

**Comment by rahulbhalley** [comment]
Transfer learning in term of S4TF API design: desired layers' writable parameters must not be moved along the gradient of model. A simple workaround is nested method I recently wrote that makes the `TangentVector` of desired layers of model zero. Now even when I update _whole_ model along all gradients these certain layers are not updated.
```swift
func zeroModelGradient(of model: inout MyNet.TangentVector) {
    func zeroLayerGradient<T: Differentiable & KeyPathIterable>(of layer: inout T) {
        for kp in layer.recursivelyAllWritableKeyPaths(to: Tensor<Float>.self) {
            layer[keyPath: kp] = Tensor(zerosLike: layer[keyPath: kp])
        }
    }
    // List all the layers to *not update*.
    zeroLayerGradient(of: &model.dense1)
}
```

This issue seems to be resolved with this nested function as a workaround. But to make the usage of `model` in `Optimizer` implementation more concrete I am leaving this issue open. 

cc @dan-zheng

**Comment by eaplatanios** [comment]
@rahulbhalley a quick answer to your question based on skimming through it is that you'd need a separate optimizer instance per "part" of the model you need to update. That is because optimizers are sometimes stateful and have internal accumulators (e.g., for the gradient moments) that correspond one-to-one to "parts" of the model being optimized. Reusing the same optimizer for different layers (even if they have the same type) is thus technically incorrect for some of the optimizers (e.g., all moment-based optimizers that keep track of moments).

**Comment by rahulbhalley** [comment]
@eaplatanios I apologize for too large issue content. And I believe you prolly misunderstood me.

In short:
> you'd need a separate optimizer instance per "part" of the model you need to update.

Above I have written two optimizers: `opt1` and `opt2` to update `myNet.dense1` and `myNet.dense12`, respectively. And each optimizer has its own separate learning rate as follows:
```swift
var lr1: Float = 0.001
var lr2: Float = 0.1
var opt1 = SGD(for: myNet.dense1, learningRate: lr1)
var opt2 = SGD(for: myNet.dense2, learningRate: lr2)
```

#### Problem/Bug
It is possible to update any layer with either `opt1` or `opt2` as long as the type of that layer is `Dense<Float>`. 

#### Why this happens?
This is because `SGD` infers the `Model` type as `Dense<Float>` when initialized with `myNet.densex` in both optimizers declared above. 

And the layer passed to `for:` argument of `init(for model: Model, ...)` of `SGD` is not used anywhere in [implementation](https://github.com/tensorflow/swift-apis/blob/7e2212efc4b41beedb40524f64649f125a554c71/Sources/TensorFlow/Optimizers/SGD.swift#L38) although it should be used in `update(...)` method for updating only that layer. 

Furthermore, `model:` argument in [`update(_ model: Model, along direction: Model.TangentVector)`](https://github.com/tensorflow/swift-apis/blob/7e2212efc4b41beedb40524f64649f125a554c71/Sources/TensorFlow/Optimizers/SGD.swift#L54) method allows updating any layer of `Model` type passed to it when called, whereas it should update only `model` passed to `SGD` during initialization.

**Comment by eaplatanios** [comment]
Oh I see what you're saying. Yes I did misunderstand. Do you have any suggestions on better ways to do this? As far as I can tell, types are the only way for the compiler to check whether you're using that function correctly. Many models are value types and so every time they are updated a new copy may be made (depending on how that model is used) and there is no way to track its identity. So, I'm not sure I would classify this as a bug.

**Comment by dan-zheng** [comment]
> #### Problem/Bug
> It is possible to update any layer with either `opt1` or `opt2` as long as the type of that layer is `Dense<Float>`.

I agree with @eaplatanios: this is actually the intended behavior.

Let `Model` be the name of an arbitrary `Layer`-conforming type. Optimizers should be able to update any `Model` given an (appropriate) instance of `Model.TangentVector`.

This design is simple and makes sense in Swift. `Model` and `Model.TangentVector` are almost always [value types](https://www.swiftbysundell.com/basics/value-and-reference-types) and don't have a referential identity. When you apply a differential operator like `let dmodel = gradient(at: model, in: body)`, there's nothing linking `dmodel` to `model`.

---

However, in practice, many `Model`s have properties with instance-specific "shapes" (e.g. tensors and arrays). It only makes sense to update a `Model` with a `Model.TangentVector` of the same shape.

For example, calling [`Differentiable.move(along:)`](https://github.com/apple/swift/blob/master/docs/DifferentiableProgramming.md#the-differentiable-protocol-1) with `model` and `gradient` of different shapes leads to a runtime error:

```swift
struct Model: Differentiable {
  var array: [Float]
}
var model = Model(array: [1, 2, 3])
let gradient = Model.TangentVector(array: [1, 2])
// Note: `model.array` has count 3 but `gradient.array` has count 2.
model.move(along: gradient)
print(model)
// Precondition failed: cannot move Array.DifferentiableView with count 3 along direction with different count 2: file /Users/swiftninjas/s4tf/swift/stdlib/public/core/../Differentiation/DifferentiationSupport.swift, line 1072
```

Statically preventing these "shape mismatch" errors isn't really possible in Swift's type system, but you shouldn't encounter this error often unless you mix up or create custom `TangentVector` instances.

**Comment by dan-zheng** [comment]
I'll close this issue now as intended behavior.

**Comment by dan-zheng** [comment]
@rahulbhalley: if you'd like to clarify specific issues you encountered, feel free to elaborate here or create a new issue!

---

Perhaps you ran into "shape mismatch" errors, as shown in https://github.com/tensorflow/swift-apis/issues/656#issuecomment-588080813. `zeroLayerGradient` in [your code snippet](https://github.com/tensorflow/swift-apis/issues/656#issuecomment-580500247) sets `TangentVector` instances to zero in a shape-compatible way.

TF-1005 tracks an easier way to accomplish this using a new `Differentiable.zeroTangentVector` instance property. It's not quite ready, but it enables the following:
```swift
func zeroModelGradient(of model: inout MyNet.TangentVector) {
    /*
    func zeroLayerGradient<T: Differentiable & KeyPathIterable>(of layer: inout T) {
        for kp in layer.recursivelyAllWritableKeyPaths(to: Tensor<Float>.self) {
            layer[keyPath: kp] = Tensor(zerosLike: layer[keyPath: kp])
        }
    }
    // List all the layers to *not update*.
    zeroLayerGradient(of: &model.dense1)
    */
    // The line below accomplishes the same as the lines above.
    model.dense1 = model.dense1.zeroTangentVector
}
```

**Comment by rahulbhalley** [comment]
Wow. Thanks @dan-zheng. Usage of `zeroTangentVector` looks pretty clean. 👍
